_posts/2016-04-01-rust-via-its-core-values.markdown

```diff
@@ -0,0 +1,343 @@
+}
+//<anon>:11:30: 11:36 error: use of moved value: `source` [E0382]
+{% endrp_highlight %}
+
+Remember the rules? Only one owner allowed. And when you reach the end of a curly brace the values are destroyed. When we call `same_length` we give it ownership of our values, and when it's done those values are destroyed. Comments make this a little easier to see.
+
+{% rp_highlight rust %}
+fn same_length(s1: String, s2: String) -> bool {
+  s1.len() == s2.len()
+}
+
+fn main() {
+  let source = String::from("Hello"); //source owns "Hello"
+  let other = String::from("Hi"); //other owns "Hi"
+  let other2 = String::from("Hola!"); //other2 owns "Hola!
+
+  println!("{}", same_length(source, other)); 
+  //We just gave `same_length` ownership of source and other, 
+  //and they were destroyed when that function ended
+
+  println!("{}", same_length(source, other2)); 
+  //source no longer owns anything
+}
+//<anon>:11:30: 11:36 error: use of moved value: `source` [E0382]
+{% endrp_highlight %}
+
+This seems pretty limiting. It's nice that Rust values memory safety so highly, but does it have to value it so highly?
+
+Rust ignores our griping and sticks to its core values by introducing Borrowing. A value can only have one owner, but any number of borrowers. Borrowing in Rust uses the `&` symbol.
+
+{% rp_highlight rust %}
+#[allow(unused_variables)]
+fn main() {
+  let original_owner = String::from("Hello");
+  let new_borrower = &original_owner;
+  println!("{}", original_owner);
+}
+//Hello
+{% endrp_highlight %}
+
+Earlier in this talk we tried to do the same thing but with transferring ownership. That failed. But with borrowing it succeeds. We can use the same approach with our function:
+
+{% rp_highlight rust %}
+fn same_length(s1: &String, s2: &String) -> bool {
+  s1.len() == s2.len()
+}
+
+fn main() {
+  let source = String::from("Hello"); //source owns "Hello"
+  let other = String::from("Hi"); //other owns "Hi"
+  let other2 = String::from("Hola!"); //other2 owns "Hola!"
+
+  println!("{}", same_length(&source, &other)); 
+  //We just borrowed source and other to same_length, so nothing is destroyed
+
+  println!("{}", same_length(&source, &other2));
+  //We can borrow source again!
+}
+//false
+//true
+{% endrp_highlight %}
+
+We explicitly borrow our data to the function, which explicitly says it only accepts borrowed things. When `same_length` finishes it stops borrowing the value, but the value is not destroyed.
+
+But wait, doesn't this break the memory safety Core Value that ownership solved? Won't code like this lead to disaster?
+
+{% rp_highlight rust %}
+fn main() {
+  let mut x = String::from("Hi!");
+  let y = &x;
+  y.truncate(0);
+  // Oh noes! Truncate deletes our string!
+}
+{% endrp_highlight %}
+
+Well, no. Rust's Core Value of memory safety leads to the following rules:
+
+- There can only be one owner
+- You can borrow a value as many times as you want, but those borrows are immutable
```

Comment:
Maybe you'd rather not go into detail, but you did leave out mutable borrowing here.